src/Ace/Halogen/Component.purs

```diff
@@ -1,75 +1,185 @@
 module Ace.Halogen.Component
   ( aceComponent
   , aceConstructor
-  , module Ace.Halogen.Component.Query
-  , module Ace.Halogen.Component.State
+  , AceEffects()
+  , Autocomplete(..)
+  , AceState(..)
+  , AceQuery(..)
+  , CompleteFn()
   ) where
 
 import Prelude
 
+import Ace.Editor as Editor
+import Ace.Ext.LanguageTools as LanguageTools
+import Ace.Ext.LanguageTools.Completer as Completer
+import Ace.Types
 import Control.Monad (when)
+import Control.Monad.Aff (Aff(), runAff)
+import Control.Monad.Aff.AVar (AVAR())
+import Control.Monad.Eff (Eff())
 import Control.Monad.Eff.Class (MonadEff)
-
-import Data.Maybe (Maybe(..))
-
-import Halogen (Natural(), Component(), ComponentHTML(), ComponentDSL(), SlotConstructor(..), component, action, liftEff', liftH, modify, gets)
+import Control.Monad.Eff.Random (random, RANDOM())
+import Control.Monad.Eff.Ref (Ref(), REF(), readRef, writeRef, modifyRef)
+import DOM (DOM())
+import DOM.HTML.Types (HTMLElement())
+import Data.Date (nowEpochMilliseconds, Now())
+import Data.Foldable (traverse_)
+import Data.Maybe (Maybe(..), maybe)
+import Data.StrMap (StrMap())
+import Data.StrMap as Sm
+import Data.Time (Milliseconds(..))
+import Halogen
 import Halogen.HTML.Indexed as H
 import Halogen.HTML.Properties.Indexed as P
 
-import Ace.Types (ACE(), Editor())
-import Ace.Editor as Editor
-import Ace.Halogen.Component.Query
-import Ace.Halogen.Component.State
+----------------------------------------------------------------------
+-- Foreign knots and effectful functions
+----------------------------------------------------------------------
+foreign import completeFns :: forall e. Ref (StrMap (CompleteFn e))
+foreign import initialized :: Ref Boolean
+foreign import focused :: Ref String
+foreign import dataset
+  :: forall e. HTMLElement -> Eff (dom :: DOM |e) (StrMap String)
```

Comment:
Should some of this stuff be defined in `purescript-ace` instead, if global things are needed to make autocomplete work?